fix(php): Add awslinux recipe to address al2023 failures

zsistla opened the pull request:
This recipe will install on al2023 in a way that complies with our stated compatibilities.

When running the Virtuoso Recipe query on status=FAILED PHP agent installs, most of them stem from the fact that yum is used to install the agent on Amazon Linux 2023, even though yum install method is not supported on Amazon Linux 2023 - https://docs.newrelic.com/docs/apm/agents/php-agent/getting-started/php-agent-compatibility-requirements/#operating-systems.

Amazon Linux 2023 (AL2023) is not directly based on a specific RHEL version. While Amazon Linux 2 was based on RHEL 7, AL2023 uses a combination of Fedora Linux and [CentOS Stream](https://www.google.com/search?sca_esv=8fe472cf0c0bad30&rlz=1C5GCCM_en&q=CentOS+Stream&sa=X&ved=2ahUKEwiatsT24KuOAxUJK0QIHSdaAWAQxccNegQIYhAB&mstk=AUtExfDcUfhfucRSOOSK0ndK_XXa4yImgzlrHFs8uZpRXa8iFaPBSZlldTNAnLQ6SwMaa6zPeohjvKum3SU-W0Zolh9RpTSLqqLYqZPKQsXym_jpybyf2-S5xfsTKmibXDKEl3DKuJFYqWDKn2o3NsDcm-zWxyV0S7qsVctfF3EcLt7YWLW543al9nFUV3vF-JOxBeaG&csui=3) as its upstream sources. Specifically, the initial release incorporates parts of Fedora 34, 35, 36, and [CentOS 9 Stream](https://www.google.com/search?sca_esv=8fe472cf0c0bad30&rlz=1C5GCCM_en&q=CentOS+9+Stream&sa=X&ved=2ahUKEwiatsT24KuOAxUJK0QIHSdaAWAQxccNegQIfRAB&mstk=AUtExfDcUfhfucRSOOSK0ndK_XXa4yImgzlrHFs8uZpRXa8iFaPBSZlldTNAnLQ6SwMaa6zPeohjvKum3SU-W0Zolh9RpTSLqqLYqZPKQsXym_jpybyf2-S5xfsTKmibXDKEl3DKuJFYqWDKn2o3NsDcm-zWxyV0S7qsVctfF3EcLt7YWLW543al9nFUV3vF-JOxBeaG&csui=3) and as such is not compatible with our rpm package.
More info here: https://support.newrelic.com/s/hubtopic/aAX8W000000L5iEWAS/php-failed-to-install-new-relic-in-my-amazon-linux-2023-server

Additionally, AL2023 uses DNF (the yum call is just a wrapper to DNF) which is "mostly" similar, but again, not compatible with our yum install method and AL2023 is specifically mentioned as a tarball only install: https://docs.newrelic.com/docs/apm/agents/php-agent/getting-started/php-agent-compatibility-requirements/#operating-systems.

The yum task was only included in the redhat.yml recipe so customers would have the option to later user the rpm manager for agent updates. 
As such the call to yum is unnecessary for AL2023 installs since only tarball installation is supported.

The query for PHP agent installs with al2023 now shows the agent getting successfully installed with the new recipe.
<img width="1451" height="322" alt="image" src="https://github.com/user-attachments/assets/acbc216b-7807-48a9-95f5-5e00743180ad" />

zsistla committed:
fix(php): Add awslinux recipe to address al2023 failures

vagrawal-newrelic commented:
Also as we are adding a new recipe specifically for Amazon linux 2023 its worth adding a new test case for PHP installation in amazon linux 2023 So that our test framework is also consistent with the new recipe added.

zsistla commented on recipes/newrelic/apm/php/awslinux.yml:
```diff
@@ -0,0 +1,563 @@
+          NR_INSTALL_USE_CP_NOT_LN=1 NR_INSTALL_SILENT=true NR_INSTALL_KEY="{{.NEW_RELIC_LICENSE_KEY}}" ./newrelic-install install
+          popd > /dev/null
+
+      vars:
```
The var block was added to prevent errors when an issue occurred finding the most current agent version (i.e., if the [curl](https://github.com/newrelic/open-install-library/pull/1256/files#diff-d6f5cb7628effe180e70040e0af81379d0f30d9f471e15335ca48347d592660aR332) command failed for any reason) so that agent version always has at least something.  This should be left as-is.

zsistla committed:
Update awslinux.yml

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/newrelic/open-install-library?pullRequest=1256) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/newrelic/open-install-library?pullRequest=1256) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/newrelic/open-install-library?pullRequest=1256) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: pranav-new-relic<br/>:x: zsistla<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/newrelic/open-install-library?pullRequest=1256) it.</sub>

zsistla commented on recipes/newrelic/apm/php/awslinux.yml:
```diff
@@ -438,8 +438,8 @@ install:
           if [ "$APPLICATION_RESTART" = "y" ] && [ -f "{{.TMP_INSTALL_DIR}}/processes_to_restart.txt" ]; then
             cd {{.TMP_INSTALL_DIR}}
             read -r process user_name < "{{.TMP_INSTALL_DIR}}/processes_to_restart.txt"
-            echo -e "{{.YELLOW}}Restarting $process as privileged user $user_name{{.GRAY}}"
```
It must be restarted by the same user that initially started it.
This call is not causing it to fail. As seen with the original recipe, it restarted the process without issue: https://github.com/newrelic/open-install-library/actions/runs/24474007706/job/71521432140#step:8:3879

The error occurs [later](https://github.com/newrelic/open-install-library/actions/runs/24474007706/job/71521432140#step:8:3884) with this message `Invalid JSON string found in output file, skipping:` and is possibly due to the test environment setup.

Here is the Dockerfile I used for testing:
```
FROM amazonlinux:2023

# Install PHP
RUN dnf update -y
#RUN dnf install -y vi tar unzip gzip
RUN dnf install -y \
	gd \
	gd-devel \
	php \
	php-fpm \
	php-common \
	php-opcache \
	php-cli \
	php-gd \
	php-curl \
	php-mysqli

# Install tools needed for virtuoso.
RUN dnf -y install \
        chkconfig \
        coreutils-single \
        gawk \
        gnupg \
        grep \
        go \
        gzip \
        procps \
        sed \
        sudo \
        tar \
        unzip \
        vim \
        which \
	&& dnf clean all

# A little setup work for php-fpm.
RUN sed -e 's/127.0.0.1:9000/9000/' \
        -e '/allowed_clients/d' \
        -e '/catch_workers_output/s/^;//' \
        -e '/error_log/d' \
        -i /etc/php-fpm.d/www.conf
RUN mkdir -p /var/www/html
EXPOSE 9003

# Copy install to run virtuoso.
COPY utils/call_virtuoso.sh /tmp/call_virtuoso.sh
RUN chmod 755 /tmp/call_virtuoso.sh

# Start php-fpm.
RUN mkdir -p /var/run/php-fpm
CMD [ "php-fpm", "-F" ]
```

When using the original recipe with the above, the recipe passes:

Installing New Relic CLI v0.111.6
Installing to /usr/local/bin

 _   _                 ____      _ _
| \ | | _____      __ |  _ \ ___| (_) ___
|  \| |/ _ \ \ /\ / / | |_) / _ | | |/ __|
| |\  |  __/\ V  V /  |  _ |  __| | | (__
|_| \_|\___| \_/\_/   |_| \_\___|_|_|\___|

Welcome to New Relic. Let's set up full stack observability for your environment.
Our Data Privacy Notice: https://newrelic.com/termsandconditions/services-notices
    
✔ Connecting to the New Relic platform
   Connected 


Installing New Relic

✔ Installing PHP Agent
   Installed 

  New Relic installation complete 

  --------------------
  What's next?
  ✔  PHP Agent  (installed)  

  View your data at the link below:
  ⮕  https://onenr.io/0Vwg7457MwJ

zsistla committed:
Update awslinux.yml

zsistla committed:
Update awslinux.yml

pranav-new-relic committed:
test(php): Add AL2023 PHP agent test definitions with minimal php-fpm deploy

- Add test definitions for US and EU regions (apache-fpm and nginx-fpm variants)
- Use minimal php-fpm deploy script instead of full WordPress stack to isolate
  the PHP agent installation from web server setup complexity
- Test definitions use local_recipes: true to pick up recipe changes from PR branch
- Validate NR CLI output for "PHP Agent (installed)" after instrumentation
- Deploy targets al2023-ami-2023.*-x86_64 with t3.micro instances

pranav-new-relic committed:
Merge branch 'main' into php/awslinux_2023

zsistla commented:
@pranav-new-relic 
I am unclear why you reverted the changes here: https://github.com/newrelic/open-install-library/commit/c65eed3fe764a11753a4165d6b09f2de01b5a94d
Which were installing wordpress on nginx and apache and then passing all tests here: https://github.com/newrelic/open-install-library/actions/runs/24487399117/job/71565260590

Adding a minimal php-fpm is a good addition, but testing cli only does not exercise the web server detection nor the restart webservice functionality since php-cli does not need a restart to get updated php changes.

With the current php-fpm only changes, [‎test/definitions/apm/us-php-nginx-fpm-wordpress-awslinux2023.json](https://github.com/newrelic/open-install-library/pull/1256/changes#diff-692f75d2a543320b69643e2872bc4b1f172fe63fd365d26f500a7487aa758760) and [‎test/definitions/apm/us-php-apache-fpm-wordpress-awslinux2023.json](https://github.com/newrelic/open-install-library/pull/1256/changes#diff-f8e730ba3b1d95ea92bc92690496b3385c0266cf4e7d0dd44ae8cb089e7be671) are both redundant since they do the exact same thing and are also misnamed since none test wordpress, nginx , apache.  It would be best to rename it `us-php-fpm-awslinux2023.json` and re-add the working apache/nginx wordpress tests from here https://github.com/newrelic/open-install-library/commit/c65eed3fe764a11753a4165d6b09f2de01b5a94d

pranav-new-relic commented:
Hey @zsistla, thank you so much for following up on my comments/commits and experimenting with the tests yesterday :) this should've helped you get an idea of the way the deployer test framework works and the definitions written as well - sorry about all of the chaos with the commits I reverted, added and squashed so far.

While a minimal test definition isn't the best thing, I had to do so in the process of untangling multiple issues which were causing failures here - issues with the test, some missing elements in the initial recipe (which you'd added yesterday) and the issue with nginx; I'll explain below.

After you'd added all the commits yesterday to fix a few things in the recipe and the tests too, while the tests were passing, I noticed changes were made in the shell script dealing in wordpress, with an `exit 0` after the restart step for nginx; given this is common to multiple tests relying on wordpress (which should not "all" be affected by this change, ideally) and the fact that we're essentially skipping nginx induced issues, I wanted to see if a minimal test definition would fix it (so we could atleast be sure the initial recipe has no issues and needs no "incorrect" modifications like I did with the restart step, which was a wrong call), which is why I pushed a commit to change the test definition to a minimal one, and without the `exit 0` change in the wordpress script - for us to be sure the recipe, as is, atleast works on AL2023 in the first place. 

This led me to discover that the install was still not working on AL2023 with this minimal definition, so I picked the changes in each commit of yours from yesterday and validate them by pushing them again to see which is causing the install to work - leading to the observation that the recipe needed changes too in https://github.com/newrelic/open-install-library/pull/1256/changes/c2e2d12b566dfcba9d8b9a696306820d1434679b and https://github.com/newrelic/open-install-library/pull/1256/changes/f57aa4276a99bf232f8da6f141c21adf383ebd81 - which is why I eventually squashed everything else with the revert, preserving the minimal definition and these two commits.

I understand this was a messy experiment, but given this resurfaced after many months, I wanted to derive a minimal test suite to see if the AL2023 install with this recipe even works in the first place, in the interest of time if you guys would like to have this merged earlier. However, I understand this might not be the real best solution, and can be given the time to address better if you guys would like; I'd like to say that this would be best your call, since you know understand the nuances of these test elements with PHP better :) if this is not a time sensitive change, kindly feel free to add any necessary changes here so we have the tests working too, and the test suite includes nginx as well - otherwise, kindly feel free to modify the minimal test suite as needed, if you find these can be named better to reflect the right state. I only wanted to chime in for a few additions to move this as we sadly kept this halted for a long time - so you could take this forward accordingly; kindly feel free to modify this as desired (I hope my observations here could help you do so). This is probably too long of a comment - thanks for reading through this :)

zsistla commented:
@pranav-new-relic 
Thanks for getting this jump started again and adding the test cases!

The real change needed for the recipe was only the order of flags on al2023 https://github.com/newrelic/open-install-library/commit/f57aa4276a99bf232f8da6f141c21adf383ebd81 but it didn't hurt to also update to systemctl so added that as well.

The nginx and apache recipes with the few additional packages were both working with the wordpress installation with the commits that were reverted: https://github.com/newrelic/open-install-library/actions/runs/24487399117/job/71565260590
The wordpress install only had 2 basic changes:
1) update all systemctl calls (which weren't causing failures but were causing redirects) but doesn't hurt to update the call.
2) "I noticed changes were made in the shell script dealing in wordpress, with an exit 0 after the restart step for nginx; given this is common to multiple tests relying on wordpress (which should not "all" be affected by this change, ideally) and the fact that we're essentially skipping nginx induced issues"

a return after the starting the nginx was added.  This is because if we've started nginx we don't need to try and start apache as well.  This will not impact any other tests using this shared script.  This change has nothing to do with wordpress and is just starting either nginx or apache, but with al2023, I saw it tried to start both nginx (which succeeded) and httpd (which failed) which led to nginx test failure before even getting to the recipe. No nginx issues are skipped, there's just really no need to try to start apache if nginx is already running.


I think the minimal php-fpm test (once renamed) is a good addition, but needs a rename.
Adding the changes again from here(https://github.com/newrelic/open-install-library/commit/c65eed3fe764a11753a4165d6b09f2de01b5a94d) to the nginx/apache/wordpress install (where both nginx and apache tests passed) will make those work tests all work once again.

zsistla committed:
rename minimal php-fpm for eu

zsistla committed:
rename us-php-fpm-awslinux2023.json to reflect php-fpm only

zsistla committed:
al2023 wordpress + lamp packages

zsistla committed:
nginx config for al2024 with php lamp packages

zsistla committed:
apache-fpm for al2023 with php lamp packages

zsistla committed:
Delete test/deploy/linux/php/apache-fpm-wordpress/aws2023/roles/configure/tasks/main.yml

zsistla committed:
modernize

* update all to latest systemctl vs service start.
* exit if nginx is successful since it can't coexist with apache anyway.

zsistla committed:
point php-apache-fpm-wordpress to apache-fpm-wordpress role

zsistla committed:
nginx-fpm-wordpress should point to nginx-fpm-wordpress

zsistla committed:
apache-fpm-wordpress should point to apache-fpm-wordpress

zsistla committed:
nginx-fpm-wordpress should point to nginx-fpm-wordpress

zsistla committed:
wrong file

zsistla committed:
update systemctl

zsistla committed:
Update install-wordpress.sh

zsistla committed:
Merge branch 'main' into php/awslinux_2023

pranav-new-relic commented:
Hey @zsistla, thank you for sharing an elaborate account :) this was really helpful. I think the changes look good to me (it took me a while to contextualize everything we discussed :)) - let me know if you'd like to double check for one final time, and we can have this merged and released.

zsistla commented:
> Hey @zsistla, thank you for sharing an elaborate account :) this was really helpful. I think the changes look good to me (it took me a while to contextualize everything we discussed :)) - let me know if you'd like to double check for one final time, and we can have this merged and released.

@pranav-new-relic Please go ahead an merge.

pranav-new-relic committed:
chore: merge main into php/awslinux_2023

pranav-new-relic commented:
LGTM, thank you for all the changes @zsistla
